File: source/c/file_watcher.c
// gcc -o file_watcher file_watcher.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>

int main(void) {
    const char *watch_path = "/tmp/hemlock_watch_test.txt";

    // Write initial file
    FILE *f = fopen(watch_path, "w");
    fputs("initial content", f); fclose(f);
    long t1 = (long)clock();
    printf("[watcher] monitoring: %s\n", watch_path);
    printf("[watcher] initial write at t=%ld\n", t1);

    // Read back
    char buf[256] = {0};
    f = fopen(watch_path, "r"); fgets(buf, sizeof(buf), f); fclose(f);
    printf("[watcher] initial content: %s\n", buf);

    // Modify
    f = fopen(watch_path, "w");
    fputs("modified content", f); fclose(f);
    long t2 = (long)clock();
    printf("[watcher] file modified at t=%ld\n", t2);

    // Detect
    memset(buf, 0, sizeof(buf));
    f = fopen(watch_path, "r"); fgets(buf, sizeof(buf), f); fclose(f);
    if (strcmp(buf, "modified content") == 0) {
        printf("[watcher] file change detected\n");
        printf("[watcher] new content: %s\n", buf);
    } else {
        printf("[watcher] no change detected\n");
    }

    if (t2 >= t1) printf("[watcher] modification time advanced\n");

    printf("[watcher] done\n");
    remove(watch_path);
    return 0;
}

File: source/python/file_watcher.py
import time
import os

watch_path = "/tmp/hemlock_watch_test.txt"

with open(watch_path, "w") as f:
    f.write("initial content")
t1 = int(time.time() * 1000)
print(f"[watcher] monitoring: {watch_path}")
print(f"[watcher] initial write at t={t1}")

with open(watch_path, "r") as f:
    initial = f.read()
print(f"[watcher] initial content: {initial}")

with open(watch_path, "w") as f:
    f.write("modified content")
t2 = int(time.time() * 1000)
print(f"[watcher] file modified at t={t2}")

with open(watch_path, "r") as f:
    content = f.read()

if content == "modified content":
    print("[watcher] file change detected")
    print(f"[watcher] new content: {content}")
else:
    print("[watcher] no change detected")

if t2 > t1:
    print("[watcher] modification time advanced")

print("[watcher] done")
